### PR TITLE
core: Apply SO_REUSEPORT to UDP sockets

### PR DESCRIPTION
**Pull request by mholt**
For some reason, 10 months ago when I implemented SO_REUSEPORT
for TCP, I didn't realize, or forgot, that it can be used for UDP too. It is a
much better solution than using deadline hacks to reuse a socket, at
least for TCP.

Then https://github.com/mholt/caddy-l4/issues/132 was posted,
in which we see that UDP servers never actually stopped when the
L4 app was stopped. I verified this using this command:

    $ nc -u 127.0.0.1 55353

combined with POSTing configs to the /load admin endpoint (which
alternated between an echo server and a proxy server so I could tell
which config was being used).

I refactored the code to use SO_REUSEPORT for UDP, but of course
we still need graceful reloads on all platforms, not just Unix, so I
also implemented a deadline hack similar to what we used for
TCP before. That implementation for TCP was not perfect, possibly
having a logical (not data) race condition; but for UDP so far it
seems to be working. Verified the same way I verified that SO_REUSEPORT
works.

I think this code is slightly cleaner and I'm fairly confident this code
is effective.

Update: The bug reporter has confirmed that this fixes the issue and does
not result in a memory leak.

### CODE REVIEW

**Comment by mholt** [comment]
@WeidiDeng If you have a chance, we might want your review to ensure HTTP/3 still works properly through config reloads on Linux and Windows. (I don't have Windows)

**Comment by WeidiDeng** [comment]
It doesn't work on windows 🤷‍♂️ . And there is a error log from [quic](https://github.com/quic-go/quic-go/blob/master/http3/server.go#L684).

image

**Review by WeidiDeng — CHANGES_REQUESTED** [review]
I added some debug output to http3 listener.

image

That error is related to `errFakeClosed` (emoji not displayed properly in powershell). This means quic listener (which is also shared) is still using the old packetconn (which is closed) instead of the new one (not closed).

**Comment by WeidiDeng** [comment]
Maybe quic listener needs something similar to the tls.Config, but then again, `fakeCloseQuicListener` already does this job well enough.

File: listen.go [hunk 1]
```diff
@@ -30,18 +30,35 @@ func reuseUnixSocket(network, addr string) (any, error) {
 	return nil, nil
 }
 
-func listenTCPOrUnix(ctx context.Context, lnKey string, network, address string, config net.ListenConfig) (net.Listener, error) {
-	sharedLn, _, err := listenerPool.LoadOrNew(lnKey, func() (Destructor, error) {
-		ln, err := config.Listen(ctx, network, address)
+func listenReusable(ctx context.Context, lnKey string, network, address string, config net.ListenConfig) (any, error) {
+	switch network {
+	case "udp", "udp4", "udp6", "unixgram":
```
[the remark this answers is not in the document]
Comment (mholt) [reply listen.go:35]: Oh, interesting. I didn't realize that. So I need to treat that as a unix socket instead.

**Comment by mholt** [comment]
Aw man.

Thanks for your tests/analysis. I'll see what I can do to fix that up. I might have more questions for you :sweat_smile:

**Comment by francislavoie** [comment]
Is this done? If not I think we should bump this to v2.7.6

**Comment by mholt** [comment]
I think I might need @WeidiDeng 's help to finish this correctly. Bumping sounds good to me!

**Comment by WeidiDeng** [comment]
I have tested this patch on linux and macos with [5749](https://github.com/caddyserver/caddy/pull/5749) and it's working fine, haven't tested it on windows though.

**Comment by mholt** [comment]
@WeidiDeng Oof, I might have butchered the rebase. Do you think you could help me verify whether I did it correctly? :pleading_face:

**Comment by mholt** [comment]
Thanks for the help again @WeidiDeng :smile: If you had a chance to try this and then review this as "approved" then I think I'll feel comfortable merging this in :+1:

**Review by WeidiDeng — APPROVED** [review]
LGTM

**Pull request merged**